src/cipher.rs

```diff
@@ -472,319 +486,651 @@ impl rustls_root_cert_store {
+    // https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs
+    _private: [u8; 0],
+}
+
+impl Castable for rustls_root_cert_store {
+    type Ownership = OwnershipArc;
+    type RustType = RootCertStore;
+}
+
+impl rustls_root_cert_store {
     /// Free a rustls_root_cert_store previously returned from rustls_root_cert_store_builder_build.
     /// Calling with NULL is fine. Must not be called twice with the same value.
     #[no_mangle]
-    pub extern "C" fn rustls_root_cert_store_free(store: *mut rustls_root_cert_store) {
+    pub extern "C" fn rustls_root_cert_store_free(store: *const rustls_root_cert_store) {
+        ffi_panic_boundary! {
+            free_arc(store);
+        }
+    }
+}
+
+/// A built client certificate verifier that can be provided to a `rustls_server_config_builder`
+/// with `rustls_server_config_builder_set_client_verifier`.
+pub struct rustls_client_cert_verifier {
+    _private: [u8; 0],
+}
+
+impl Castable for rustls_client_cert_verifier {
+    type Ownership = OwnershipBox;
+    type RustType = Arc<dyn ClientCertVerifier>;
+}
+
+impl rustls_client_cert_verifier {
+    /// Free a `rustls_client_cert_verifier` previously returned from
+    /// `rustls_client_cert_verifier_builder_build`. Calling with NULL is fine. Must not be
+    /// called twice with the same value.
+    #[no_mangle]
+    pub extern "C" fn rustls_client_cert_verifier_free(verifier: *mut rustls_client_cert_verifier) {
         ffi_panic_boundary! {
-            let store = try_box_from_ptr!(store);
-            drop(store)
+            free_box(verifier);
         }
     }
 }
 
-/// A builder for a `rustls_allow_any_authenticated_client_verifier`. This builder object can be
-/// used to configure certificate revocation lists, and then turned into a
-/// `rustls_allow_any_authenticated_client_verifier` once ready.
-pub struct rustls_allow_any_authenticated_client_builder {
+/// A client certificate verifier being constructed. A builder can be modified by,
+/// e.g. `rustls_web_pki_client_cert_verifier_builder_add_crl`. Once you're
+/// done configuring settings, call `rustls_web_pki_client_cert_verifier_builder_build`
+/// to turn it into a `rustls_client_cert_verifier`. This object is not safe
+/// for concurrent mutation.
+// TODO(@cpu): Add rustdoc link once available.
+pub struct rustls_web_pki_client_cert_verifier_builder {
+    // We use the opaque struct pattern to tell C about our types without
+    // telling them what's inside.
+    // https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs
     _private: [u8; 0],
 }
 
-impl Castable for rustls_allow_any_authenticated_client_builder {
+pub(crate) struct ClientCertVerifierBuilder {
+    roots: Arc<RootCertStore>,
+    root_hint_subjects: Vec<DistinguishedName>,
+    crls: Vec<CertificateRevocationListDer<'static>>,
+    revocation_depth: RevocationCheckDepth,
+    revocation_policy: UnknownStatusPolicy,
+    allow_anonymous: bool,
```

Comment:
```suggestion
    allow_unauthenticated: bool,
```

(and corresponding changes anywhere this field is referenced)

Reply:
Done!